### PR TITLE
Add separate message for filtered all stream

### PR DESCRIPTION
**Pull request by Yspadadden**
Removed the previous solution with an additional argument to ReadAllEventsForward. 

Added a new command ReadAllEventsForwardFiltered with its own semantic. 

Also now supporting regexes in the allowed filters.

### CODE REVIEW

File: src/EventStore.ClientAPI/AllEventsFilteredSlice.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿using EventStore.ClientAPI.Messages;
+
+namespace EventStore.ClientAPI
+{
+    public class AllEventsFilteredSlice : AllEventsSlice
+    {
+        private readonly bool IsEndOfStreamInternal;
```
Comment (daMupfel) [inline src/EventStore.ClientAPI/AllEventsFilteredSlice.cs]: Is the internal property really needed? I have no idea myself ;)

File: src/EventStore.ClientAPI/AllEventsFilteredSlice.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿using EventStore.ClientAPI.Messages;
```
Comment (daMupfel) [inline src/EventStore.ClientAPI/AllEventsFilteredSlice.cs]: Add this file to EventStore.ClientAPI/EventStore.ClientAPI.csproj:
EventStore.ClientAPI/EventStore.ClientAPI.csproj

File: src/EventStore.ClientAPI/AllEventsFilteredSlice.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace EventStore.ClientAPI
+{
+    public class AllEventsFilteredSlice : AllEventsSlice
```
Comment (daMupfel) [inline src/EventStore.ClientAPI/AllEventsFilteredSlice.cs]: Add comments to your class. (See AllEventsSlice as example)

File: src/EventStore.ClientAPI/AllEventsFilteredSlice.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    {
+        private readonly bool IsEndOfStreamInternal;
+
+        public override bool IsEndOfStream { get { return IsEndOfStreamInternal; } }
```
Comment (daMupfel) [inline src/EventStore.ClientAPI/AllEventsFilteredSlice.cs]: I am not sure why this even works?

https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/override

To my knowledge you can only override abstract, virtual or already overriden methods / members

File: src/EventStore.Core.Tests/Services/Storage/AllReader/when_reading_all_with_filtering.cs [hunk 5]
```diff
@@ -23,7 +23,7 @@ public void should_read_only_allowed_events()
         {
```
Comment (daMupfel) [inline src/EventStore.Core.Tests/Services/Storage/AllReader/when_reading_all_with_filtering.cs:23]: Not part of this pull request but it looks like the style is here not to use this (see WriteTestScenario())

File: src/EventStore.Core.Tests/Services/Storage/AllReader/when_reading_all_with_filtering.cs [hunk 6]
```diff
@@ -23,7 +23,7 @@ public void should_read_only_allowed_events()
...
             var expectedEventTypes = new StringFilter(new string[] { "event-type" });
```
Comment (daMupfel) [inline src/EventStore.Core.Tests/Services/Storage/AllReader/when_reading_all_with_filtering.cs]:
```suggestion
            var expectedEventTypes = new StringFilter(new[] { "event-type" });
```

File: src/EventStore.Core.Tests/Util/StringFilterTest.cs [hunk 7]
```diff
@@ -76,5 +76,48 @@ public void substrings_do_not_match_for_single_string()
             Assert.False(sf.IsStringAllowed("the"));
             Assert.False(sf.IsStringAllowed("hedge"));
         }
+
+        [Test]
+        public void regex_supported_in_match()
+        {
+            StringFilter sf = new StringFilter(new string[] { "foo-.+" });
```
Comment (daMupfel) [inline src/EventStore.Core.Tests/Util/StringFilterTest.cs]: Use var:
var sf = new StringFilter(new string[] { "foo-.+" });

File: src/EventStore.Core/Messages/ClientMessage.cs [hunk 8]
```diff
@@ -759,7 +757,7 @@ public class ReadAllEventsForwardCompleted : ReadResponseMessage
             public readonly TFPos PrevPos;
             public readonly long TfLastCommitPosition;
 
-            public bool IsEndOfStream { get { return Events == null || Events.Length < MaxCount; } }
+            public virtual bool IsEndOfStream { get { return Events == null || Events.Length < MaxCount; } }
```
Comment (daMupfel) [inline src/EventStore.Core/Messages/ClientMessage.cs]: Maybe i am blind, but who is overriding this? I don't see anyone inheriting from ReadAllEventsForwardCompleted

File: src/EventStore.Core/Services/Storage/ReaderIndex/AllReader.cs [hunk 9]
```diff
@@ -17,7 +17,13 @@ public interface IAllReader
         /// Returns event records in the sequence they were committed into TF.
         /// Positions is specified as pre-positions (pointer at the beginning of the record).
         /// </summary>
-        IndexReadAllResult ReadAllEventsForward(TFPos pos, int maxCount, StringFilter allowedEventTypes);
+        IndexReadAllResult ReadAllEventsForward(TFPos pos, int maxCount);
+
+        /// <summary>
+        /// Returns event records in the sequence they were committed into TF.
```
Comment (daMupfel) [inline src/EventStore.Core/Services/Storage/ReaderIndex/AllReader.cs]: Add to comment why this method is different then ReadAlLEVentsForward (i know that you kind of can deduce from the name and parameters). Like this it is kinda pointless.

File: src/EventStore.Core/Services/Storage/ReaderIndex/AllReader.cs [hunk 10]
```diff
@@ -57,12 +91,10 @@ public IndexReadAllResult ReadAllEventsForward(TFPos pos, int maxCount, StringFi
             long consideredEventsCount = 0;
```
Comment (daMupfel) [inline src/EventStore.Core/Services/Storage/ReaderIndex/AllReader.cs]:
```suggestion
            long consideredEventsCount = 0L;
```
Otherwise you are converting an int to long

Comment (daMupfel) [inline src/EventStore.Core/Services/Storage/ReaderIndex/AllReader.cs]: You can use var with some of the values here. I am not sure it makes the code more clear though.
For all assignments where the type is clear i usually prefer var.
Example:
bool firstCommit = true; vs var firstCommit = true;

File: src/EventStore.Core/Services/Storage/ReaderIndex/AllReader.cs [hunk 11]
```diff
@@ -43,11 +48,40 @@ public AllReader(IIndexBackend backend, IIndexCommitter indexCommitter, ICheckpo
...
+
+                Records = records;
+                CurrentPos = currentPos;
+                NextPos = nextPos;
+                PrevPos = prevPos;
+                IsEndOfStream = isEndOfStream;
+            }
+        }
+
+        private ReadAllEventsForwardRawResult ReadAllEventsForwardInternal(TFPos pos, int maxCount, int maxSearchWindow, StringFilter allowedEventTypes)
```
Comment (daMupfel) [inline src/EventStore.Core/Services/Storage/ReaderIndex/AllReader.cs:83]: Like last time it is hard for  me to see if this i 100% correct. Just from looking at it, it seems to do what we want.

File: src/EventStore.Core/Services/Storage/ReaderIndex/IReadIndex.cs [hunk 12]
```diff
@@ -22,12 +22,12 @@ public interface IReadIndex
         /// Positions is specified as pre-positions (pointer at the beginning of the record).
         /// </summary>
         IndexReadAllResult ReadAllEventsForward(TFPos pos, int maxCount);
-        IndexReadAllResult ReadAllEventsForward(TFPos pos, int maxCount, StringFilter allowedEventTypes);
         /// <summary>
         /// Returns event records in the reverse sequence they were committed into TF.
         /// Positions is specified as post-positions (pointer after the end of record).
         /// </summary>
         IndexReadAllResult ReadAllEventsBackward(TFPos pos, int maxCount);
+        IndexReadAllFilteredResult ReadAllEventsForwardFiltered(TFPos pos, int maxCount, int maxSearchWindow, StringFilter allowedEventTypes);
```
Comment (daMupfel) [inline src/EventStore.Core/Services/Storage/ReaderIndex/IReadIndex.cs:34]: Add comment, looks like this would be the same comment as the other place has.

File: src/EventStore.Core/Services/Storage/ReaderIndex/IndexReadAllFilteredResult.cs [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+            Records = records;
+            CurrentPos = currentPos;
+            NextPos = nextPos;
+            PrevPos = prevPos;
+            IsEndOfStream = isEndOfStream;
+        }
+
+        public override string ToString()
+        {
+            return string.Format("CurrentPos: {0}, NextPos: {1}, PrevPos: {2}, IsEndOfStream: {3}, Records: {4}",
```
Comment (daMupfel) [inline src/EventStore.Core/Services/Storage/ReaderIndex/IndexReadAllFilteredResult.cs:29]: Use string interpolation:
```
            var records = string.Join("\n", Records.Select(x => x.ToString()));
            return $"CurrentPos: {CurrentPos}, NextPos: {NextPos}, PrevPos: {PrevPos}, IsEndOfStream: {IsEndOfStream}, Records: {records}";
```

Comment (daMupfel) [reply src/EventStore.Core/Services/Storage/ReaderIndex/IndexReadAllFilteredResult.cs:29]: Hm, the rest of the code seems to use string.Format(). So i think that should be fine.

File: src/EventStore.Core/Services/Storage/StorageReaderWorker.cs [hunk 14]
```diff
@@ -148,6 +149,48 @@ void IHandle<ClientMessage.ReadAllEventsBackward>.Handle(ClientMessage.ReadAllEv
             msg.Envelope.ReplyWith(ReadAllEventsBackward(msg));
         }
 
+        void IHandle<ClientMessage.ReadAllEventsForwardFiltered>.Handle(ClientMessage.ReadAllEventsForwardFiltered msg)
+        {
+            if (msg.Expires < DateTime.UtcNow)
+            {
+                Log.Debug("Read All Stream Events Forward operation has expired for C:{0}/P:{1}. Operation Expired at {2}", msg.CommitPosition, msg.PreparePosition, msg.Expires);
```
Comment (daMupfel) [inline src/EventStore.Core/Services/Storage/StorageReaderWorker.cs]: This looks like the "Read All Stream Events Forward Filtered" operation?

File: src/EventStore.Core/Util/StringFilter.cs [hunk 15]
```diff
@@ -88,5 +114,54 @@ public bool IsStringAllowed(string s)
...
+            private readonly List<IFilterStrategy> strategies;
+
+            public MultiStrategyStrategy(List<IFilterStrategy> strategies)
+            {
+                this.strategies = strategies;
+            }
+
+            public bool IsStringAllowed(string s)
+            {
+                foreach(var strat in strategies)
```
Comment (daMupfel) [inline src/EventStore.Core/Util/StringFilter.cs:156]: Please rename to strategy.
https://www.urbandictionary.com/define.php?term=strats

Comment (daMupfel) [reply src/EventStore.Core/Util/StringFilter.cs:156]: I am not sure about performance implications but this could also be written like this:
`return strategies.Exists(strategy => strategy.IsStringAllowed(s));`

File: src/EventStore.ClientAPI/AllEventsFilteredSlice.cs (hunk 1 above)
Comment (daMupfel) [reply src/EventStore.ClientAPI/AllEventsFilteredSlice.cs]: This should work (but no guarantees):
``` csharp
    /// <summary>
    /// Comment
    /// </summary>
    public class AllEventsFilteredSlice : AllEventsSlice
    {
        /// <summary>
        /// Comment
        /// </summary>
        public override bool IsEndOfStream { get; }

        internal AllEventsFilteredSlice(ReadDirection readDirection, Position fromPosition, Position nextPosition, ClientMessage.ResolvedEvent[] events, bool isEndOfStream) 
            : base(readDirection, fromPosition, nextPosition, events)
        {
            IsEndOfStream = isEndOfStream;
        }
    }
```

File: src/Protos/ClientAPI/ClientMessageDtos.proto [hunk 16]
```diff
@@ -197,6 +196,36 @@ message ReadAllEventsCompleted {
...
+    required int32 max_count = 3;
+    required bool resolve_link_tos = 4;
+    required bool require_master = 5;
+    optional int32 max_search_window = 6;
+    repeated string allowed_event_types = 7;
+}
+
+message ReadAllEventsFilteredCompleted {
+
+    enum ReadAllResult {
```
Comment (daMupfel) [inline src/Protos/ClientAPI/ClientMessageDtos.proto]: Should probably be called ReadAllFilteredResult

File: src/Protos/ClientAPI/ClientMessageDtos.proto [hunk 17]
```diff
@@ -197,6 +196,36 @@ message ReadAllEventsCompleted {
 	optional string error = 7;
 }
 
+message ReadAllEventsFiltered {
+    required int64 commit_position = 1;
+    required int64 prepare_position = 2;
+    required int32 max_count = 3;
+    required bool resolve_link_tos = 4;
+    required bool require_master = 5;
+    optional int32 max_search_window = 6;
```
Comment (daMupfel) [inline src/Protos/ClientAPI/ClientMessageDtos.proto]: Wouldn't it make sense for this to be at position 4?
The message does not need to be binary compatible with ReadAlLEvents anyway.

**Review by daMupfel — COMMENTED** [review]
Review done

File: src/EventStore.Core/Messages/ClientMessage.cs (hunk 8 above)
Comment (Yspadadden) [reply src/EventStore.Core/Messages/ClientMessage.cs]: Not anymore. I first tried to subclass this. But this doesn't work, because the wrong handler will be called. Will remove.

File: src/EventStore.Core/Services/Storage/ReaderIndex/IndexReadAllFilteredResult.cs (hunk 13 above)
Comment (Yspadadden) [reply src/EventStore.Core/Services/Storage/ReaderIndex/IndexReadAllFilteredResult.cs:29]: So, should I leaf it as is then?

File: src/EventStore.Core/Util/StringFilter.cs (hunk 15 above)
Comment (Yspadadden) [reply src/EventStore.Core/Util/StringFilter.cs:156]: First impression is that this would make the code less clear. `Exists` sounds to me like we expect something to exist. But it is just a filter. I am probably not c# developer enough to like this.

File: src/EventStore.Core/Services/Storage/ReaderIndex/IndexReadAllFilteredResult.cs (hunk 13 above)
Comment (daMupfel) [reply src/EventStore.Core/Services/Storage/ReaderIndex/IndexReadAllFilteredResult.cs:29]: yes, you can :four_leaf_clover: it

File: src/EventStore.Core/Util/StringFilter.cs (hunk 15 above)
Comment (daMupfel) [reply src/EventStore.Core/Util/StringFilter.cs:156]: I am fine with both. For me Exists looks c#y

Comment (daMupfel) [reply src/EventStore.Core/Util/StringFilter.cs:156]: Or use Any, which is an extension method for Enumerables.
``` csharp
    public static bool Any<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate)
    {
      if (source == null)
        throw Error.ArgumentNull(nameof (source));
      if (predicate == null)
        throw Error.ArgumentNull(nameof (predicate));
      foreach (TSource source1 in source)
      {
        if (predicate(source1))
          return true;
      }
      return false;
    }
```

Comment (daMupfel) [reply src/EventStore.Core/Util/StringFilter.cs:156]: Or just leave it as is!

Comment (Yspadadden) [reply src/EventStore.Core/Util/StringFilter.cs:156]: I'll leave it as is then.

**Review by daMupfel — APPROVED** [review]

**Review by daMupfel — APPROVED** [review]

**Review by daMupfel — CHANGES_REQUESTED** [review]
No Approve

**Review by daMupfel — APPROVED** [review]
ReApprove
